Adding build script to build repo on linux

ahsonkhan opened the pull request:
cc @joshfree, @shiftylogic, @KrzysztofCwalina 

I added a build.sh in attempts to build corefxlab repo on linux. The install-dotnet.sh script is from https://github.com/dotnet/cli/blob/master/scripts/obtain/dotnet-install.sh.

**Here are the issues I found:** Edit: All issues resolved.
1) ~**Edit:** Filed an issue to track the issue outside of this PR: #1536. Using the new roslyn compiler leads to no project being built successfully on linux. Is this a known issue (@VSadov, any ideas)? We need to find a way to resolve this. To move forward, I reverted to the compiler that comes with the cli (i.e. removed https://github.com/dotnet/corefxlab/blob/master/tools/common.props#L17-L19). **Assuming we resolve this issue, we have the issues that follow.**
Example error:
`/home/aak/.nuget/packages/microsoft.net.compilers/2.3.0-rdonly-ref-61607-04/tools/Microsoft.CSharp.Core.targets(71,5): error MSB3883: Unexpected exception:  [/home/aak/git/corefxlab/src/System.Text.Primitives/System.Text.Primitives.csproj]
/home/aak/.nuget/packages/microsoft.net.compilers/2.3.0-rdonly-ref-61607-04/tools/Microsoft.CSharp.Core.targets(71,5): error : System.ComponentModel.Win32Exception (0x80004005): Permission denied [/home/aak/git/corefxlab/src/System.Text.Primitives/System.Text.Primitives.csproj]
/home/aak/.nuget/packages/microsoft.net.compilers/2.3.0-rdonly-ref-61607-04/tools/Microsoft.CSharp.Core.targets(71,5): error :    at Interop.Sys.ForkAndExecProcess(String filename, String[] argv, String[] envp, String cwd, Boolean redirectStdin, Boolean redirectStdout, Boolean redirectStderr, Int32& lpChildPid, Int32& stdinFd, Int32& stdoutFd, Int32& stderrFd) [/home/aak/git/corefxlab/src/System.Text.Primitives/System.Text.Primitives.csproj]
/home/aak/.nuget/packages/microsoft.net.compilers/2.3.0-rdonly-ref-61607-04/tools/Microsoft.CSharp.Core.targets(71,5): error :    at System.Diagnostics.Process.StartCore(ProcessStartInfo startInfo) [/home/aak/git/corefxlab/src/System.Text.Primitives/System.Text.Primitives.csproj]
/home/aak/.nuget/packages/microsoft.net.compilers/2.3.0-rdonly-ref-61607-04/tools/Microsoft.CSharp.Core.targets(71,5): error :    at System.Diagnostics.Process.Start() [/home/aak/git/corefxlab/src/System.Text.Primitives/System.Text.Primitives.csproj]
/home/aak/.nuget/packages/microsoft.net.compilers/2.3.0-rdonly-ref-61607-04/tools/Microsoft.CSharp.Core.targets(71,5): error :    at Microsoft.Build.Utilities.ToolTask.ExecuteTool(String pathToTool, String responseFileCommands, String commandLineCommands) [/home/aak/git/corefxlab/src/System.Text.Primitives/System.Text.Primitives.csproj]
/home/aak/.nuget/packages/microsoft.net.compilers/2.3.0-rdonly-ref-61607-04/tools/Microsoft.CSharp.Core.targets(71,5): error :    at Microsoft.CodeAnalysis.BuildTasks.ManagedCompiler.ExecuteTool(String pathToTool, String responseFileCommands, String commandLineCommands) [/home/aak/git/corefxlab/src/System.Text.Primitives/System.Text.Primitives.csproj]`~
2) ~For some reason package restore (+build/test) fails for System.Binary.Base64.tests.csproj. I am not sure why we get the error message:
`/home/aak/git/corefxlab/dotnet/sdk/2.0.0-preview2-005905/NuGet.targets(715,5): warning MSB3202: The project file "/home/aak/git/corefxlab/tests/System.Binary.Base64.Tests/System.Binary.Base64.tests.csproj" was not found. [/home/aak/git/corefxlab/corefxlab.sln]`
It is there (@shiftylogic, any ideas?): https://github.com/dotnet/corefxlab/blob/master/corefxlab.sln#L95~
3) ~The following projects fail to build on Linux:~
- ~System.CommandLine.csproj~
- ~System.Time.csproj~
~`/home/aak/git/corefxlab/dotnet/sdk/2.0.0-preview2-005905/Microsoft.Common.CurrentVersion.targets(1111,5): error MSB3644: The reference assemblies for framework ".NETFramework,Version=v4.5" were not found. To resolve this, install the SDK or Targeting Pack for this framework version or retarget your application to a version of the framework for which you have the SDK or Targeting Pack installed. Note that assemblies will be resolved from the Global Assembly Cache (GAC) and will be used in place of reference assemblies. Therefore your assembly may not be correctly targeted for the framework you intend. [/home/aak/git/corefxlab/src/System.CommandLine/System.CommandLine.csproj]
/home/aak/git/corefxlab/dotnet/sdk/2.0.0-preview2-005905/Microsoft.Common.CurrentVersion.targets(1111,5): error MSB3644: The reference assemblies for framework ".NETFramework,Version=v4.5" were not found. To resolve this, install the SDK or Targeting Pack for this framework version or retarget your application to a version of the framework for which you have the SDK or Targeting Pack installed. Note that assemblies will be resolved from the Global Assembly Cache (GAC) and will be used in place of reference assemblies. Therefore your assembly may not be correctly targeted for the framework you intend. [/home/aak/git/corefxlab/src/System.Time/System.Time.csproj]`~
4) ~The following tests fail on Linux. For now, should we disable these tests for any platforms other than windows (@KrzysztofCwalina, any ideas)?
System.Time.Tests.DateTimeOffsetTests.CanAddYearsAcrossDstTransition
System.Time.Tests.DateTimeOffsetTests.CanAddDaysAcrossDstTransition_StartWithMismatchedOffset
System.Time.Tests.DateTimeOffsetTests.CanAddDaysAcrossDstTransition_LandInOverlap
System.Time.Tests.DateTimeOffsetTests.CanAddHoursAcrossDstTransition
System.Time.Tests.DateTimeOffsetTests.CanAddTicksAcrossDstTransition
System.Time.Tests.DateTimeOffsetTests.CanAddDaysAcrossDstTransition
System.Time.Tests.DateTimeOffsetTests.CanAddMonthsAcrossDstTransition
System.Time.Tests.DateTimeOffsetTests.CanAddMillisecondsAcrossDstTransition
System.Time.Tests.DateTimeOffsetTests.CanAddDaysAcrossDstTransition_LandInGap
System.Time.Tests.DateTimeOffsetTests.CanAddSecondsAcrossDstTransition
System.Time.Tests.DateTimeOffsetTests.CanAddMinutesAcrossDstTransition
System.Time.Tests.DateTimeTests.CanAddMinutesAcrossDstTransition
System.Time.Tests.DateTimeTests.CanAddMillisecondsAcrossDstTransition
System.Time.Tests.DateTimeTests.CanAddHoursAcrossDstTransition_StartWithMismatchedKind
System.Time.Tests.DateTimeTests.CanAddDaysAcrossDstTransition_LandInGap
System.Time.Tests.DateTimeTests.CanAddHoursAcrossDstTransition
System.Time.Tests.DateTimeTests.CanAddDaysAcrossDstTransition
System.Time.Tests.DateTimeTests.CanAddTicksAcrossDstTransition
System.Time.Tests.DateTimeTests.CanAddYearsAcrossDstTransition
System.Time.Tests.DateTimeTests.CanAddDaysAcrossDstTransition_LandInOverlap
System.Time.Tests.DateTimeTests.CanAddMonthsAcrossDstTransition
System.Time.Tests.DateTimeTests.CanAddDaysAcrossDstTransition_StartWithMismatchedKind
System.Time.Tests.DateTimeTests.CanAddSecondsAcrossDstTransition
System.Slices.Tests.HttpRequestTests.HttpReaderMultipleSegments
System.Slices.Tests.HttpRequestTests.HttpReaderSingleSegment
PollingWatcherUnitTests.FileSystemWatcher_Deleted_File
PollingWatcherUnitTests.FileSystemWatcher_Created_File
PollingWatcherUnitTests.FileSystemWatcher_Changed_File
Example failed test errors:
a) `System.TimeZoneNotFoundException : The time zone ID 'Pacific Standard Time' was not found on the local computer.
---- System.IO.FileNotFoundException : Could not find file '/usr/share/zoneinfo/Pacific Standard Time'.`
b) `System.InvalidOperationException : Nullable object must have a value.`
c) `System.DllNotFoundException : Unable to load DLL 'kernel32.dll': The specified module or one of its dependencies could not be found.
 (Exception from HRESULT: 0x8007007E)`~
5) ~[Minor] I was testing on Ubuntu 17.04 and installing dotnet cli was failing (which I believe is expected since CLI only has a package for Ubuntu 16.04). I manually installed [dotnet cli](https://github.com/dotnet/cli) (for Ubuntu 16.04) to move forward (@livarcocc, am I using an unsupported version of ubuntu?).
`dotnet_install: Error: OS name could not be detected: ubuntu.17.04`~

joshfree commented:
> System.TimeZoneNotFoundException

You'll want to conditionally use America/Los_Angeles on Linux for that test case.

ahsonkhan commented:
- Renamed System.Binary.Base64.tests.csproj to System.Binary.Base64.Tests.csproj (resolves issue 2)
- Removed leftover net45 tfm in System.CommandLine.csproj / System.Time.csproj (resolves issue 3)
- Disabled tests in SocketsFacts.cs for linux (since they were hanging, opened issue #1533 - cc @pakrym, @davidfowl)
- Updated to a newer compiler that should work for linux as well. Disabling the use of that compiler for linux (issue 1 still open). I am working with @VSadov to try to resolve this issue.
- Using ubuntu 16.04 (resolves issue 5)
- For issue 4:
  - Resolved System.Time.Tests failures by applying the suggestion from @joshfree
  - Disabled System.IO.FileSystem.Watcher tests since PollingWatcher uses [Win32 API](https://github.com/dotnet/corefxlab/blob/master/src/System.IO.FileSystem.Watcher.Polling/System/IO/PollingWatcher.cs#L95).
  - Fixed HttpRequest tests by fixing Parse implementation to work for both Windows and Linux (carriage return is /r/n in Windows while just /n for Linux).
 - Added HasValue checks for Nullable type in HttpRequest Parse

davidfowl commented on tools/common.props:
```diff
@@ -14,7 +14,7 @@
     <AssemblyOriginatorKeyFile>$(MSBuildThisFileDirectory)Key.snk</AssemblyOriginatorKeyFile>
   </PropertyGroup>
-  <ItemGroup>
+  <ItemGroup Condition=" '$(OS)' == 'Windows_NT' ">
```
There's a linux version /cc @VSadov @agocke

ahsonkhan replied:
Yes, I know... except:
> Updated to a newer compiler that should work for linux as well. Disabling the use of that compiler for linux (issue 1 still open). I am working with @VSadov to try to resolve this issue.

ahsonkhan replied:
Filed issue #1536

shiftylogic commented on src/System.Text.Http/System/Text/Http/HttpRequest.cs:
```diff
@@ -22,8 +22,14 @@ public override string ToString()
+#if Windows
         internal static readonly byte[] Cr = new byte[] { (byte)'\r', (byte)'\n' };
         internal static readonly byte[] Cr2 = new byte[] { (byte)'\r', (byte)'\n', (byte)'\r', (byte)'\n' };
+#else
```
This is the wrong fix. The fix should be done in the test sources to force a CRLF in the pre-fab HTTP requests since the HTTP spec requires CRLF.

shiftylogic commented on src/System.Text.Http/System/Text/Http/HttpRequest.cs:
```diff
@@ -47,21 +53,28 @@ public static HttpRequest Parse(ReadOnlyBytes bytes)
             var reader = new BytesReader(bytes);
             Range? verb = reader.ReadRangeUntil((byte)' ');
-            request._verb = verb.Value;
+            if (verb.HasValue) request._verb = verb.Value;
```
Once you fix the tests, these shouldn't be needed anymore.

shiftylogic commented on tests/System.Time.Tests/DateTimeOffsetTests.cs:
```diff
@@ -7,10 +7,17 @@ namespace System.Time.Tests
+#if Windows
+    string TimeZoneId = "Pacific Standard Time";
+#else
+    string TimeZoneId = "America/Los_Angeles";
```
You should try running these tests on Mac. Are these strings Linux specific, or UNIX (and therefore Mac). Otherwise, you might need more change here to cover Mac.

ahsonkhan replied:
These strings should work for Linux and Mac. Still, I will see if I can run the tests on Mac as well.

https://stackoverflow.com/tags/timezone/info

> The IANA/Olson Time Zone Database
Widely implemented on Linux, Mac, Java, PHP, and many other platforms.

shiftylogic commented on tools/common.props:
```diff
@@ -14,7 +14,4 @@
     <AssemblyOriginatorKeyFile>$(MSBuildThisFileDirectory)Key.snk</AssemblyOriginatorKeyFile>
   </PropertyGroup>
-  <ItemGroup>
-    <PackageReference Include="Microsoft.Net.Compilers" Version="$(RoslynVersion)" PrivateAssets="All" />
```
We aren't using the new compiler at all now?

ahsonkhan replied:
Opened issue #1536. We were not using the right compiler to begin with (it should be Microsoft.NetCore.Compilers). That package was built against framework 1.1.0 and hence using it on a machine that doesn't have 1.1.0 and only has the dotnet cli that we bring down does not work.

ahsonkhan commented:
@joshfree, good to go?
